File: go/luhn/luhn.go
package luhn

import (
	"regexp"
	"strconv"
	"strings"
)

// Valid determines whether or not the input is valid per the Luhn formula.
func Valid(ccnumber string) bool {
	ccRegExp := regexp.MustCompile(`^\d[\d ]+$`)

	if !ccRegExp.MatchString(ccnumber) {
		return false
	}

	ccnumberArray := strings.Split(strings.Replace(ccnumber, " ", "", -1), "")
	nDigits := len(ccnumberArray)
	parity := nDigits % 2
	sum, _ := strconv.Atoi(ccnumberArray[nDigits-1])

	for i := 0; i < nDigits-1; i++ {
		digit, _ := strconv.Atoi(ccnumberArray[i])
		if i%2 == parity {
			digit *= 2
		}
		if digit > 9 {
			digit -= 9
		}
		sum += digit
	}

	return (sum % 10) == 0
}

File: javascript/luhn/luhn.js
//@ts-check

/**
 * @param {string} ccnumber
 * @returns {boolean}
 */
export const valid = ccnumber => {
  if (!/^\d[\d ]+$/.test(ccnumber)) {
    return false;
  }

  const result = [...ccnumber.replace(/ /g, '')].map(digit => parseInt(digit, 10));

  const digitsCount = result.length;
  const parity = digitsCount % 2;
  let sum = result[digitsCount - 1];

  for (let i = 0; i < digitsCount - 1; i++) {
    let digit = result[i];
    if (i % 2 === parity) {
      digit *= 2;
    }
    if (digit > 9) {
      digit = digit - 9;
    }
    sum += digit;
  }

  return sum % 10 === 0;
};
